File: Version1/Templates/MVC Module.xctemplate/___FILEBASENAME___/Controller/___FILEBASENAME___ViewController.m
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//___COPYRIGHT___
//

#import "___VARIABLE_moduleName___ViewController.h"
#import "___VARIABLE_moduleName___Model.h"
#import "___VARIABLE_moduleName___View.h"


@interface ___VARIABLE_moduleName___ViewController () <___VARIABLE_moduleName___ModelDelegate, ___VARIABLE_moduleName___ViewDelegate>

/** @brief A model */
@property (nonatomic, strong) NSObject <___VARIABLE_moduleName___ModelProtocol> *model;

/** @brief Overriding view property */
@property (nonatomic, strong) UIView<___VARIABLE_moduleName___ViewProtocol> *view;

@property (nonatomic, strong) ___VARIABLE_moduleName___Router *router;

@end


@implementation ___VARIABLE_moduleName___ViewController

@dynamic view;
@dynamic model;
@dynamic router;


#pragma mark - Initialization

- (nullable id)initWithView:(nonnull UIView<___VARIABLE_moduleName___ViewProtocol> *)view model:(nonnull NSObject<___VARIABLE_moduleName___ModelProtocol> *)model router:(nullable ___VARIABLE_moduleName___Router *)router {
    self = [super initWithNibName:NSStringFromClass([self class]) bundle:nil view:view model:model];
    if (self) {
        self.model.delegate = self;
        self.router = router;
        
        // custom initialization code goes here
    }
    return self;
}


#pragma mark - View lifecycle

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // assign a view delegate when the view did load
    self.view.delegate = self;
    
    // custom code goes here
}
    
        
#pragma mark - ___VARIABLE_moduleName___ViewDelegate
    
/** Implement ___VARIABLE_moduleName___ViewDelegate methods here */

- (void)viewSomeAction:(UIView<___VARIABLE_moduleName___ViewProtocol> *)view {
    NSLog(@"Something happened on view!");
}


#pragma mark - ___VARIABLE_moduleName___ModelDelegate

- (void)modelDidChanged:(NSObject<___VARIABLE_moduleName___ModelProtocol> *)model {
    
}

@end

File: Version3/Templates/MVC V3 Swift Module.xctemplate/___FILEBASENAME___/___FILEBASENAME___ViewController.swift
//
//  ___FILENAME___
//  ___PROJECTNAME___
//
//  Created by ___FULLUSERNAME___ on ___DATE___.
//  ___COPYRIGHT___
//

import UIKit

// MARK: - ___VARIABLE_moduleName___ViewController

class ___VARIABLE_moduleName___ViewController: UIViewController {

    var model: ___VARIABLE_moduleName___ModelProtocol

    fileprivate var tempView: ___VARIABLE_moduleName___ViewProtocol?
    var customView: ___VARIABLE_moduleName___ViewProtocol! {
        return self.view as? ___VARIABLE_moduleName___ViewProtocol
    }

    // MARK: Initializers

    init(withView view: ___VARIABLE_moduleName___ViewProtocol, model: ___VARIABLE_moduleName___ModelProtocol) {
        self.model = model
        self.tempView = view

        super.init(nibName: nil, bundle: nil)
    }

    required convenience init?(coder aDecoder: NSCoder) {
        fatalError("This class needs to be initialized with init(withView:model:) method")
    }

    // MARK: - View life cycle

    override func loadView() {
        super.loadView()

        self.view = self.tempView
        self.tempView = nil
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        customView.delegate = self
        model.delegate = self
    }

    // MARK: - Private methods

}

// MARK: - ___VARIABLE_moduleName___ViewDelegate

extension ___VARIABLE_moduleName___ViewController: ___VARIABLE_moduleName___ViewDelegate {

    func viewSomeAction(view: ___VARIABLE_moduleName___ViewProtocol) {
    }
}

// MARK: - ___VARIABLE_moduleName___ModelDelegate

extension ___VARIABLE_moduleName___ViewController: ___VARIABLE_moduleName___ModelDelegate {

}
